src/mds/StrayManager.cc

```diff
@@ -0,0 +1,863 @@
+    ops_avail = max_purge_ops - ops_in_flight;
+  } else {
+    ops_avail = 0;
+  }
+
+  /* The ops_in_flight > 0 condition here handles the case where the
+   * ops required by this inode would never fit in the limit: we wait
+   * instead until nothing else is running */
+  if (ops_in_flight > 0 && ops_avail < ops_required) {
+    dout(20) << __func__ << ": throttling on max ops (require "
+             << ops_required << ", " << ops_in_flight << " in flight" << dendl;
+    return false;
+  }
+
+  // Resources are available, acquire them and execute the purge
+  files_purging += 1;
+  dout(10) << __func__ << ": allocating allowance "
+    << ops_required << " to " << ops_in_flight << " in flight" << dendl;
+  ops_in_flight += ops_required;
+  logger->set(l_mdc_num_purge_ops, ops_in_flight);
+  if (trunc) {
+    truncate(dn, ops_required);
+  } else {
+    purge(dn, ops_required);
+  }
+  return true;
+}
+
+uint32_t StrayManager::_calculate_ops_required(CInode *in, bool trunc)
+{
+  uint32_t ops_required = 0;
+  if (in->is_dir()) {
+    // Directory, count dirfrags to be deleted
+    std::list<frag_t> ls;
+    if (!in->dirfragtree.is_leaf(frag_t())) {
+      in->dirfragtree.get_leaves(ls);
+    }
+    // One for the root, plus any leaves
+    ops_required = 1 + ls.size();
+  } else {
+    // File, work out concurrent Filer::purge deletes
+    const uint64_t period = (uint64_t)in->inode.layout.fl_object_size *
+		      (uint64_t)in->inode.layout.fl_stripe_count;
+    const uint64_t to = MAX(in->inode.max_size_ever,
+            MAX(in->inode.size, in->inode.get_max_size()));
+
+    const uint64_t num = MAX(1, (to + period - 1) / period);
+    ops_required = MIN(num, g_conf->filer_max_purge_ops);
+
+    // Account for removing (or zeroing) backtrace
+    ops_required += 1;
+
+    // Account for deletions for old pools
+    if (!trunc) {
+      ops_required += in->get_projected_inode()->old_pools.size();
+    }
+  }
+
+  return ops_required;
+}
+
+void StrayManager::advance_delayed()
+{
+  for (elist<CDentry*>::iterator p = delayed_eval_stray.begin(); !p.end(); ) {
+    CDentry *dn = *p;
+    ++p;
+    dn->item_stray.remove_myself();
+    num_strays_delayed--;
+
+    if (dn->get_projected_linkage()->is_null()) {
+      /* A special case: a stray dentry can go null if its inode is being
+       * re-linked into another MDS's stray dir during a shutdown migration. */
+      dout(4) << __func__ << ": delayed dentry is now null: " << *dn << dendl;
+      continue;
+    }
+
+    const bool purging = eval_stray(dn);
+    if (!purging) {
+      derr << "Dentry " << *dn << " was purgeable but no longer is!" << dendl;
+      /*
+       * This can happen if a stray is purgeable, but has gained an extra
+       * reference by virtue of having its backtrace updated.
+       * FIXME perhaps we could simplify this further by
+       * avoiding writing the backtrace of purge-ready strays, so
+       * that this code could be more rigid?
```

Comment:
The dentries in this path haven't been through enqueue yet (these are the ones eval_stray hit with delay=true).  We could put a clear_dirty_parent in that path as well (and then assert here) but I'd like to test that separately